src/main/java/org/codehaus/mojo/versions/api/AbstractVersionDetails.java

```diff
@@ -284,25 +269,13 @@ public final ArtifactVersion[] getVersions( Restriction restriction, boolean inc
     public final ArtifactVersion[] getVersions( VersionRange versionRange, Restriction restriction,
                                                 boolean includeSnapshots )
     {
-        final VersionComparator versionComparator = getVersionComparator();
-        Set<ArtifactVersion> result = new TreeSet<>( versionComparator );
-        for ( ArtifactVersion candidate : getVersions( includeSnapshots ) )
-        {
-            if ( versionRange != null && !ArtifactVersions.isVersionInRange( candidate, versionRange ) )
-            {
-                continue;
-            }
-            if ( restriction != null && !isVersionInRestriction( restriction, candidate ) )
-            {
-                continue;
-            }
-            if ( !includeSnapshots && ArtifactUtils.isSnapshot( candidate.toString() ) )
-            {
-                continue;
-            }
-            result.add( candidate );
-        }
-        return result.toArray( new ArtifactVersion[0] );
+        return Arrays.stream( getVersions( includeSnapshots ) )
+                .filter( candidate ->
+                        ( versionRange == null || ArtifactVersions.isVersionInRange( candidate, versionRange ) )
+                        && ( restriction == null || isVersionInRestriction( restriction, candidate ) )
+                        && ( includeSnapshots || !ArtifactUtils.isSnapshot( candidate.toString() ) ) )
+                .collect( () -> new TreeSet<>( getVersionComparator() ), TreeSet::add, Set::addAll )
```

Comment:
Can be worked, something like:

```
.sorted( versionComparator )
.distinct()
.collect( toList() )
```

it be more clear what we want to achieve

Reply:
I thought TreeSet has better performance, but it doesn't really matter in this case 😄